src/Razor/src/Microsoft.AspNetCore.Razor.Language/DefaultRazorTagHelperBinderPhase.cs

```diff
@@ -188,5 +256,159 @@ private bool AssemblyContainsTagHelpers(string assemblyName, IReadOnlyList<TagHe
                 return false;
             }
         }
+
+        internal class ComponentDirectiveVisitor : DirectiveVisitor
+        {
+            private IReadOnlyList<TagHelperDescriptor> _tagHelpers;
+            private string _filePath;
+            private RazorSourceDocument _source;
+
+            public ComponentDirectiveVisitor(string filePath, IReadOnlyList<TagHelperDescriptor> tagHelpers, string currentNamespace)
+            {
+                _filePath = filePath;
+
+                // We don't want to consider non-component tag helpers in a component document.
+                _tagHelpers = tagHelpers.Where(t => IsComponentTagHelperKind(t.Kind)).ToList();
+
+                for (var i = 0; i < _tagHelpers.Count; i++)
+                {
+                    var tagHelper = _tagHelpers[i];
+                    if (tagHelper.IsComponentFullyQualifiedNameMatch())
+                    {
+                        // If the component descriptor matches for a fully qualified name, using directives shouldn't matter.
+                        Matches.Add(tagHelper);
+                    }
+
+                    if (currentNamespace != null && IsTypeInScope(tagHelper.GetTypeName(), currentNamespace))
+                    {
+                        // Also, if the type is already in scope of the document's namespace, using isn't necessary.
+                        Matches.Add(tagHelper);
+                    }
+                }
+            }
+
+            public override HashSet<TagHelperDescriptor> Matches { get; } = new HashSet<TagHelperDescriptor>();
+
+            public override string TagHelperPrefix => null;
+
+            public override void Visit(RazorSyntaxTree tree)
+            {
+                _source = tree.Source;
+                Visit(tree.Root);
+            }
+
+            public override void VisitRazorDirective(RazorDirectiveSyntax node)
+            {
+                var descendantLiterals = node.DescendantNodes();
+                foreach (var child in descendantLiterals)
+                {
+                    if (!(child is CSharpStatementLiteralSyntax literal))
+                    {
+                        continue;
+                    }
+
+                    var context = literal.GetSpanContext();
+                    if (context == null)
+                    {
+                        // We can't find a chunk generator.
+                        continue;
+                    }
+                    else if (context.ChunkGenerator is AddTagHelperChunkGenerator addTagHelper)
+                    {
+                        // Make sure this node exists in the file we're parsing and not in its imports.
+                        if (_filePath.Equals(_source.FilePath, StringComparison.Ordinal))
```

Comment:
note: this condition may be the best we can do right now, but long term I suspect this check won't be correct. We'd eventually like to have the system be dumber about stuff like this.

Reply:
> long term I suspect this check won't be correct

What functionality here do you expect will change in the long term? This check is need when we introduce `_Imports.razor`. If someone puts `@addTagHelper` in `_Imports.razor`, we don't want that error to show up in every document that uses that imports. It should only show up when they open `_Imports.razor`.

Reply:
@rynowak ping

Reply:
Oh, I think this is fine for now.